ostd/src/arch/x86/tdx_guest.rs

```diff
@@ -54,21 +58,11 @@ pub unsafe fn unprotect_gpa_range(gpa: Paddr, page_num: usize) -> Result<(), Pag
             priv_flags: prop.priv_flags | PrivFlags::SHARED,
         }
     };
-    let _ = boot_pt::with_borrow(|boot_pt| {
-        for i in 0..page_num {
-            let vaddr = paddr_to_vaddr(gpa + i * PAGE_SIZE);
-            // SAFETY: The caller ensures that the address range exists in the linear mapping and
-            // can be mapped as shared pages.
-            unsafe { boot_pt.protect_base_page(vaddr, protect_op) };
-        }
-    });
 
     // Protect the page in the kernel page table.
     let pt = KERNEL_PAGE_TABLE.get().unwrap();
-    let vaddr = paddr_to_vaddr(gpa);
-    // SAFETY: The caller ensures that the address range exists in the linear mapping and can be
-    // mapped as shared pages.
-    unsafe { pt.protect_flush_tlb(&(vaddr..vaddr + page_num * PAGE_SIZE), protect_op) }
+    // SAFETY: The caller ensures that the address range can be mapped as shared pages.
+    unsafe { pt.protect_flush_tlb(&(gva..gva + page_num * PAGE_SIZE), protect_op) }
```

Comment:
I think both `unprotect_gpa_range` and `protect_gpa_range` methods should remove the call to `protect_flush_tlb`. There are two users of the two methods: `IoMem` and `DmaXxx`. In both cases, before calling `unprotect_gpa_range`, they manipulate the page table, creating new mappings or updating protection bits. Either way, they already have the chance to add the SHARED bit. Calling `protect_flush_tlb` here, simply to set the SHARED bit, doubles the work. To fix this inefficiency, remove the call to `protect_flush_tlb` from the two functions and let their callers to set the SHARE bit properly when they manipulate the page table.

Reply:
> I agree. But currently implementing it for DMA without #2351 would look like a wrong-to-wrong fix, since it would force the DMA implementation to protect the linear mapping when it does not have to.

Thanks for the insight. I will need to squeeze the time to review #2351 ...